### PR TITLE
WIP: Make commas optional in a few more language constructs.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Hrmm.  Can't seem to mark this as currently a work in progress.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 1]
```diff
@@ -1025,9 +1020,17 @@ internal AttributeArgumentListSyntax ParseAttributeArgumentList()
                                 {
                                     break;
                                 }
-                                else if (this.CurrentToken.Kind == SyntaxKind.CommaToken || this.IsPossibleAttributeArgument())
+                                else if (this.CurrentToken.Kind == SyntaxKind.CommaToken ||
+                                         this.IsPossibleAttributeArgument())
```
Comment (AlenPelin) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: This is code style change. Wouldn't it be more appropriate to move that to separate commit?

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 2]
```diff
@@ -1761,16 +1764,22 @@ private BaseListSyntax ParseBaseList()
...
                 {
-                    if (this.CurrentToken.Kind == SyntaxKind.OpenBraceToken ||
-                        this.IsCurrentTokenWhereOfConstraintClause())
+                    if (AtEndOfBaseList())
                     {
                         break;
                     }
-                    else if (this.CurrentToken.Kind == SyntaxKind.CommaToken || this.IsPossibleType())
+                    else if (this.CurrentToken.Kind == SyntaxKind.CommaToken ||
+                             this.IsPossibleType())
```
Comment (AlenPelin) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: This is code style change. Wouldn't it be more appropriate to move that to separate commit?

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 3]
```diff
@@ -3662,9 +3677,17 @@ private static bool CanReuseBracketedParameterList(CSharp.Syntax.BracketedParame
                             {
                                 break;
                             }
-                            else if (this.CurrentToken.Kind == SyntaxKind.CommaToken || this.IsPossibleParameter())
+                            else if (this.CurrentToken.Kind == SyntaxKind.CommaToken ||
+                                     this.IsPossibleParameter())
```
Comment (AlenPelin) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: This is code style change. Wouldn't it be more appropriate to move that to separate commit?

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 4]
```diff
@@ -4983,13 +5006,20 @@ private TypeParameterListSyntax ParseTypeParameterList()
...
+                    if (AtEndOfTypeParameterList())
                     {
                         break;
                     }
                     else if (this.CurrentToken.Kind == SyntaxKind.CommaToken)
                     {
                         parameters.AddSeparator(this.EatToken(SyntaxKind.CommaToken));
+
+                        // Check for optional trailing comma.
+                        if (AtEndOfTypeParameterList())
```
Comment (AlenPelin) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Wouldn't it be better to rename to `IsAtEndOfTypeParameterList`?

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 5]
```diff
@@ -5384,9 +5420,16 @@ private void ParseTypeArgumentList(out SyntaxToken open, SeparatedSyntaxListBuil
                 {
                     break;
                 }
-                else if (this.CurrentToken.Kind == SyntaxKind.CommaToken || this.IsPossibleType())
+                else if (this.CurrentToken.Kind == SyntaxKind.CommaToken ||
+                         this.IsPossibleType())
```
Comment (AlenPelin) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: This is code style change. Wouldn't it be more appropriate to move that to separate commit?

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 6]
```diff
@@ -9510,17 +9553,22 @@ internal BracketedArgumentListSyntax ParseBracketedArgumentList()
...
-                            if (this.CurrentToken.Kind == SyntaxKind.CloseParenToken ||
-                                this.CurrentToken.Kind == SyntaxKind.CloseBracketToken ||
-                                this.CurrentToken.Kind == SyntaxKind.SemicolonToken)
+                            if (AtEndOfArgumentList())
                             {
                                 break;
                             }
-                            else if (this.CurrentToken.Kind == SyntaxKind.CommaToken || this.IsPossibleArgumentExpression())
+                            else if (this.CurrentToken.Kind == SyntaxKind.CommaToken ||
+                                     this.IsPossibleArgumentExpression())
```
Comment (AlenPelin) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: This is code style change. Wouldn't it be more appropriate to move that to separate commit?

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 7]
```diff
@@ -10791,7 +10846,7 @@ private ParameterListSyntax ParseLambdaParameterList()
 
                         // additional parameters
                         int tokenProgress = -1;
-                        while(IsMakingProgress(ref tokenProgress))
+                        while (IsMakingProgress(ref tokenProgress))
```
Comment (AlenPelin) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: This is code style change. Wouldn't it be more appropriate to move that to separate commit?

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 8]
```diff
@@ -9510,17 +9553,22 @@ internal BracketedArgumentListSyntax ParseBracketedArgumentList()
...
                                 break;
                             }
-                            else if (this.CurrentToken.Kind == SyntaxKind.CommaToken || this.IsPossibleArgumentExpression())
+                            else if (this.CurrentToken.Kind == SyntaxKind.CommaToken ||
+                                     this.IsPossibleArgumentExpression())
                             {
+                                // Check for optional trailing comma
                                 list.AddSeparator(this.EatToken(SyntaxKind.CommaToken));
+
+                                if (AtEndOfArgumentList())
```
Comment (AlenPelin) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Wouldn't it be better to rename to `IsAtEndOfArgumentList`?

**Comment by CyrusNajmabadi** [comment]
The code was easier to write and validate when the patterns and style were in sync across all of them.  We do loop parsing in many different ways, and this was an attempt to make things simpler and easier to understand by bringing consistency.  Not all style changes are innapropriate.  For a codebase with lots of duplication (and an aversion to encapsulating this logic for perf reasons), style is actually important.

File: src/Compilers/CSharp/Test/Syntax/Parsing/SeparatedSyntaxListParsingTests.cs [hunk 9]
```diff
@@ -90,28 +87,26 @@ class C
...
                         N(SyntaxKind.VariableDeclaration);
                         N(SyntaxKind.GenericName);
                         N(SyntaxKind.IdentifierToken);
                         N(SyntaxKind.TypeArgumentList);
                         N(SyntaxKind.LessThanToken);
-                        {
+                        { 
                             N(SyntaxKind.IdentifierName); N(SyntaxKind.IdentifierToken);
-                            N(SyntaxKind.CommaToken);
+                            M(SyntaxKind.CommaToken);
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Syntax/Parsing/SeparatedSyntaxListParsingTests.cs]: this is more correct. the comma is actually missing.

File: src/Compilers/CSharp/Test/Syntax/Parsing/SeparatedSyntaxListParsingTests.cs [hunk 10]
```diff
@@ -158,16 +150,15 @@ class C
                         N(SyntaxKind.TypeArgumentList);
                         N(SyntaxKind.LessThanToken);
                         {
-                            N(SyntaxKind.IdentifierName); N(SyntaxKind.IdentifierToken);
+                            M(SyntaxKind.IdentifierName); M(SyntaxKind.IdentifierToken);
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Syntax/Parsing/SeparatedSyntaxListParsingTests.cs]: this is more correct.  the identifier is actually missing.
